### PR TITLE
Fixed #17209 -- Added password change/reset class-based views

### PR DESCRIPTION
**Pull request by claudep**

### CODE REVIEW

File: django/contrib/auth/views.py [hunk 1]
```diff
@@ -224,6 +224,9 @@ def password_reset(request,
                    extra_context=None,
                    html_email_template_name=None,
                    extra_email_context=None):
+    warnings.warn("Function-based views for password reset are deprecated. "
```
Comment (timgraham) [inline django/contrib/auth/views.py]: could you follow the message style  of login/logout? e.g. 'The logout() view is superseded by the class-based LogoutView().'

File: django/contrib/auth/views.py [hunk 2]
```diff
@@ -334,6 +346,117 @@ def password_reset_complete(request,
     return TemplateResponse(request, template_name, context)
 
 
+# New class-based password reset views
```
Comment (timgraham) [inline django/contrib/auth/views.py]: I'd chop "new" so we don't have to remove it later

File: docs/releases/1.11.txt [hunk 3]
```diff
@@ -70,6 +70,14 @@ Minor features
   :class:`~django.contrib.auth.views.LogoutView` class-based views supersede the
   deprecated ``login()`` and ``logout()`` function-based views.
 
+* The :class:`~django.contrib.auth.views.PasswordResetView`,
+  :class:`~django.contrib.auth.views.PasswordResetDoneView`,
+  :class:`~django.contrib.auth.views.PasswordResetConfirmView` and
```
Comment (timgraham) [inline docs/releases/1.11.txt]: , and

File: docs/releases/1.11.txt [hunk 4]
```diff
@@ -70,6 +70,14 @@ Minor features
...
+  :class:`~django.contrib.auth.views.PasswordResetCompleteView` class-based
+  views supersede the deprecated ``password_reset()``,
+  ``password_reset_done()``, ``password_reset_confirm()`` and
```
Comment (timgraham) [inline docs/releases/1.11.txt]: , and

File: docs/releases/1.11.txt [hunk 5]
```diff
@@ -70,6 +70,14 @@ Minor features
...
+  ``password_reset_complete`` function-based views.
```
Comment (timgraham) [inline docs/releases/1.11.txt]: ()

File: django/contrib/auth/views.py [hunk 6]
```diff
@@ -334,6 +346,117 @@ def password_reset_complete(request,
...
+        if self.extra_context is not None:
+            context.update(self.extra_context)
+        return context
+
+
+class PasswordResetView(PasswordContextMixin, FormView):
+    template_name = 'registration/password_reset_form.html'
+    form_class = PasswordResetForm
+    success_url = reverse_lazy('password_reset_done')
+
```
Comment (timgraham) [inline django/contrib/auth/views.py]: I see the reasoning for the line break but the arguments in the other views are organized similarly. Maybe one group arranged alphabetically would be simplest?

File: django/contrib/auth/views.py [hunk 7]
```diff
@@ -334,6 +346,117 @@ def password_reset_complete(request,
...
+        form.save()
+        return super(PasswordResetConfirmView, self).form_valid(form)
+
+    def get_context_data(self, **kwargs):
+        context = super(PasswordResetConfirmView, self).get_context_data(**kwargs)
+        user = context['form'].user
+        if user is not None and self.token_generator.check_token(user, self.kwargs['token']):
+            context['validlink'] = True
+        else:
+            del context['form']
```
Comment (timgraham) [inline django/contrib/auth/views.py]: Not sure it makes a difference but before it looks like we got `form=None` in the context.

**Comment by timgraham** [comment]
There's a "Using the views" section in topics/auth/default.html that still references the function-based views (password_change).

**Comment by claudep** [comment]
The `password_change` reference in "Using the views" refers to the URL name, AFAICS. Did I miss anything?

File: docs/internals/deprecation.txt [hunk 8]
```diff
@@ -15,7 +15,9 @@ about each item can often be found in the release notes of two versions prior.
 See the :ref:`Django 1.11 release notes<deprecated-features-1.11>` for more
 details on these changes.
 
-* ``contrib.auth.views.login()`` and ``logout()`` will be removed.
+* ``contrib.auth.views.login()``, ``logout()``, ``password_change()``,
+  ``password_change_done()``, ``password_reset()``, ``password_reset_done()``,
+  ``password_reset_confirm()`` and ``password_reset_complete`` will be removed.
```
Comment (timgraham) [inline docs/internals/deprecation.txt]: , and
missing () on  password_reset_complete

File: docs/releases/1.11.txt [hunk 9]
```diff
@@ -321,3 +332,14 @@ Miscellaneous
   deprecated in favor of new class-based views
   :class:`~django.contrib.auth.views.LoginView` and
   :class:`~django.contrib.auth.views.LogoutView`.
+
+* ``contrib.auth``’s ``password_change()``, ``password_change_done()``,
+  ``password_reset()``, ``password_reset_done()``, ``password_reset_confirm()``
```
Comment (timgraham) [inline docs/releases/1.11.txt]: , and

File: docs/releases/1.11.txt [hunk 10]
```diff
@@ -321,3 +332,14 @@ Miscellaneous
...
+  and ``password_reset_complete`` function-based views are deprecated in favor
```
Comment (timgraham) [inline docs/releases/1.11.txt]: missing ()

File: docs/topics/auth/default.txt [hunk 11]
```diff
@@ -1193,65 +1181,118 @@ implementation details see :ref:`using-the-views`.
...
 .. function:: password_change(request, template_name='registration/password_change_form.html', post_change_redirect=None, password_change_form=PasswordChangeForm, current_app=None, extra_context=None)
 
-    Allows a user to change their password.
+    .. deprecated:: 1.11
+
+        The ``password_change`` function-based view should be replaced by the
+        class-based :class:`PasswordChangeView`.
+
+    The optional arguments of this view are similar to the class-based
+    ``PasswordChangeView`` optional attributes, except the
```
Comment (timgraham) [inline docs/topics/auth/default.txt]: Missing this in the first PR for login/logout, but chop "optional"?

**Comment by timgraham** [comment]
With all the edits for the first commit in the second commit, I guess you plan to squash the commits?

File: docs/topics/auth/default.txt [hunk 12]
```diff
@@ -1360,48 +1390,83 @@ implementation details see :ref:`using-the-views`.
...
 
     **URL name:** ``password_reset_confirm``
 
-    **Optional arguments:**
+    Presents a form for entering a new password.
+
+    **Attributes:**
 
-    * ``uidb64``: The user's id encoded in base 64. Defaults to ``None``.
+    * ``uidb64``: The user's id encoded in base 64.
```
Comment (timgraham) [inline docs/topics/auth/default.txt]: Neither this nor token are attributes.

**Comment by timgraham** [comment]
Have we lost all tests for the deprecated views? I thought that was less risky for login/logout which just call the class-based views but I'm not so happy about for the other views.

**Comment by claudep** [comment]
> With all the edits for the first commit in the second commit, I guess you plan to squash the commits?

Do you see some values with separated commits (that was my first intent)?

**Comment by timgraham** [comment]
I see two commits in the PR. I only reviewed the first one the other day and all the updates for that feedback looks like it's in the second commit.

File: docs/topics/auth/default.txt [hunk 13]
```diff
@@ -1193,65 +1181,115 @@ implementation details see :ref:`using-the-views`.
...
 .. function:: password_change(request, template_name='registration/password_change_form.html', post_change_redirect=None, password_change_form=PasswordChangeForm, current_app=None, extra_context=None)
 
-    Allows a user to change their password.
+    .. deprecated:: 1.11
+
+        The ``password_change`` function-based view should be replaced by the
+        class-based :class:`PasswordChangeView`.
+
+    The optional arguments of this view are similar to the class-based
+    ``PasswordChangeView`` attributes, except the ``post_change_redirect`` and ``password_change_form`` arguments which map to the ``success_url`` and
```
Comment (timgraham) [inline docs/topics/auth/default.txt]: line length

File: tests/auth_tests/urls_deprecated.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
+from django.conf.urls import url
+from django.contrib import admin
+from django.contrib.auth import views
+from django.contrib.auth.decorators import login_required
+
+
+# special urls for deprecated function-based views
+urlpatterns = [
+    url(r'^login/$', views.LoginView.as_view(), name='login'),
```
Comment (timgraham) [inline tests/auth_tests/urls_deprecated.py]: views.login/logout

**Pull request merged**